cin_validator/rules/cin2022_23/rule_8720.py

```diff
@@ -0,0 +1,110 @@
+from typing import Mapping
+
+import pandas as pd
+
+from cin_validator.rule_engine import (
+    CINTable,
+    IssueLocator,
+    RuleContext,
+    rule_definition,
+)
+from cin_validator.test_engine import run_rule
+from cin_validator.utils import make_census_period
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+# Replace ChildIdentifiers with the table name, and LAChildID with the column name you want.
+
+ChildProtectionPlans = CINTable.ChildProtectionPlans
+CPPstartDate = ChildProtectionPlans.CPPstartDate
+Header = CINTable.Header
+ReferenceDate = Header.ReferenceDate
+
+# define characteristics of rule
+@rule_definition(
+    # write the rule code here, in place of 8720
+    code=8720,
+    # replace ChildIdentifiers with the value in the module column of the excel sheet corresponding to this rule .
+    module=CINTable.ChildProtectionPlans,
+    # replace the message with the corresponding value for this rule, gotten from the excel sheet.
+    message="Child Protection Plan Start Date missing or out of data collection period",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[CPPstartDate, ReferenceDate],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # Replace ChildIdentifiers with the name of the table you need.
+    df = data_container[ChildProtectionPlans]
+
+    # ReferenceDate exists in the heder table so we get header table too.
+    df_ref = data_container[Header]
+    ref_date_series = df_ref[ReferenceDate]
+
+    # the make_census_period function generates the start and end date so that you don't have to do it each time.
+    collection_start, reference_date = make_census_period(ref_date_series)
+
+    # implement rule logic as described by the Github issue. Put the description as a comment above the implementation as shown.
+
+    # Where a CPP module is present, <CPPstartDate> (N00105) must be present and on or before the <ReferenceDate> (N00603)
+    failing_indices = df[
+        (df[CPPstartDate] < collection_start) | (df[CPPstartDate] > reference_date)
+    ].index
+
```

Comment:
Hi @amyrichards0n , I'm not quite sure this logic works for this rule. This looks like it's checking if CPPstartDate is within the census period, where 8720 is looking at, if there is a CP plan, the plan must start before the reference date. 

I think the confusion comes from the example linked on the issue page which is for a rule that checks if something is within the census period.

I think your rule wants to check if there's a CPPID, which tells us if there is a CPP module, and if there is, if there's a CP plan start date, and it also wants to check if CP plan start dates are before the reference period.

Does this sound right @tab1tha ?

Reply:
Having a look ...

Reply:
Hi @amyrichards0n, the rule logic
> Where a CPP module is present, <CPPstartDate> (N00105) must be present and on or before the <ReferenceDate> (N00603)

can be broken into these two parts. Flag locations where

1. CPP module is present and CPPstartDate is not present. [`CPPID.notna() & CPPstartDate.isna()`]
2. CPPstartDate is present but it is greater than ReferenceDate. [CPPstartDate.notna() & (CPPstartDate > ReferenceDate)]

As Will said, this rule does need you to get the reference date from the Header table. However, since it doesn't require you to check that the date exists within the census period, you don't have to use the make_census_period function. You can go ahead and compare using the ReferenceDate as it is.

Sidenote. You might also notice that the condition `[CPPstartDate.notna() & (CPPstartDate > ReferenceDate)]` can be replaced with `[CPPstartDate > ReferenceDate]` because this check ignores NaNs by default. That is, if either of the dates are missing, the result of the comparison of dates will be NaN and so that position will not be picked by the filter.

Does this help?.